### ISSUE TITLE
Does not work with routing to the same action

### ISSUE THREAD

**Issue by davidyell** [World-Architects/cakephp-query-param-preserver#7]
## What I did
```php
Router::prefix('admin', function (RouteBuilder $routes) {
    $routes->connect('/content/static-pages', ['controller' => 'Contents', 'action' => 'index', 'type' => 1]);
    $routes->connect('/content/articles', ['controller' => 'Contents', 'action' => 'index', 'type' => 3]);
    $routes->connect('/content/our-reviews', ['controller' => 'Contents', 'action' => 'index', 'type' => 6]);

    $routes->fallbacks();
}
```

## What I expected to happen
The page would load fine, and preserve params for that route.

## What actually happened
Whenever I hit one of the urls, such as `example.com/admin/content/articles` I would be redirected to `example.com/admin/contents?content_type_id=3`.

## Workaround
I can only think that the component would need to be enhanced to understand the routing and store it's preserved params with the route it was loaded on.

**Comment by burzum** [comment]
@davidyell could you provide an unit test that shows the problem?

**Comment by davidyell** [comment]
Great, thanks for the solution @burzum I'll give that a try.

**Comment by burzum** [comment]
@davidyell could you do a pull request if it works fine for you? :)

**Comment by davidyell** [comment]
Well I don't think a pull request will really work, I've had to re-work a significant portion of the component to get things working.

```php
<?php
namespace App\Controller\Component;

use Psa\QueryParamPreserver\Controller\Component\QueryParamPreserverComponent;

/**
 * PreserveFilter component
 */
class PreserveFilterComponent extends QueryParamPreserverComponent
{
    /**
     * Builds the hash key for the current call
     *
     * @return array Hash key
     */
    protected function _hashKey(): array
    {
        $request = $this->getController()->request;
        $route = $request->getParam('_matchedRoute');

        return [$route => $request->getQuery()];
    }

    /**
     * Preserves the current query params
     *
     * @return void
     */
    public function preserve()
    {
        $request = $this->getController()->request;
        $query = $request->getQuery();

        if ($query) {
            $ignoreParams = $this->getConfig('ignoreParams');
            if (!empty($ignoreParams)) {
                foreach ($ignoreParams as $param) {
                    if (isset($query[$param])) {
                        unset($query[$param]);
                    }
                }
            }

            $this->getController()->request->getSession()->write($this->_hashKey());
        }
    }

    /**
     * Applies the preserved query params
     *
     * @return \Cake\Http\Response|null
     */
    public function apply()
    {
        $request = $this->getController()->request;
        $currentUrl = $request->getUri()->getPath();

        if (empty($request->getQuery()) && $request->getSession()->check($currentUrl)) {
            if(!empty($request->getSession()->read($currentUrl))) {
                return $this->getController()->redirect(
                    $currentUrl
                    . '?' . http_build_query($request->getSession()->read($currentUrl)));
            }
        }
    }

    /**
     * beforeFilter callback
     *
     * @return \Cake\Http\Response|null
     */
    public function beforeFilter()
    {
        $request = $this->getController()->request;
        $params = $request->getQueryParams();
        $ignoreParam = $this->getConfig('disablePreserveWithParam');

        if ($this->getConfig('autoApply') && $this->actionCheck()) {
            if (isset($params[$ignoreParam])) {
                unset($params[$ignoreParam]);

                $request->getSession()->delete($request->getUri()->getPath());

                return $this->getController()->redirect($request->getUri()->getPath());
            }

            return $this->apply();
        }
    }
}

```

**Comment by davidyell** [comment]
Well I'm not sure a PR will work here, as I've ended up having to change quite a bit of the component to get it working. In order to get it to work I've had to use the route as a string, which then meant I had to change a few more methods as well. Hopefully there is something helpfull.

```php
<?php
namespace App\Controller\Component;

use Psa\QueryParamPreserver\Controller\Component\QueryParamPreserverComponent;

/**
 * PreserveFilter component
 */
class PreserveFilterComponent extends QueryParamPreserverComponent
{
    /**
     * Builds the hash key for the current call
     *
     * @return string Hash key
     */
    protected function _hashKey(): string
    {
        return $this->getController()->request->getParam('_matchedRoute');
    }

    /**
     * Preserves the current query params
     *
     * @return void
     */
    public function preserve(): void
    {
        $request = $this->getController()->request;
        $query = $request->getQuery();

        if ($query) {
            $ignoreParams = $this->getConfig('ignoreParams');
            if (!empty($ignoreParams)) {
                foreach ($ignoreParams as $param) {
                    if (isset($query[$param])) {
                        unset($query[$param]);
                    }
                }
            }

            $this->getController()->request->getSession()->write([
                $this->_hashKey() => $query
            ]);
        }
    }

    /**
     * Applies the preserved query params
     *
     * @return \Cake\Http\Response|null
     */
    public function apply()
    {
        $request = $this->getController()->request;
        $key = $this->_hashKey();

        if (empty($request->getQuery()) && $request->getSession()->check($key)) {
            if(!empty($request->getSession()->read($key))) {
                return $this->getController()->redirect(
                    $key
                    . '?' . http_build_query($request->getSession()->read($key)));
            }
        }
    }

    /**
     * beforeFilter callback
     *
     * @return \Cake\Http\Response|null
     */
    public function beforeFilter()
    {
        $request = $this->getController()->request;
        $params = $request->getQueryParams();
        $ignoreParam = $this->getConfig('disablePreserveWithParam');

        if ($this->getConfig('autoApply') && $this->actionCheck()) {
            if (isset($params[$ignoreParam])) {
                unset($params[$ignoreParam]);

                $request->getSession()->delete($this->_hashKey());

                return $this->getController()->redirect($this->_hashKey());
            }

            return $this->apply();
        }
    }
}
```

**Comment by burzum** [comment]
@davidyell thank you but a PR will be always better because I can simply see the difference in it :)

**Comment by davidyell** [comment]
Okay, if you're sure. I'll fork, push and submit a pr tomorrow.

**Comment by davidyell** [comment]
I'll close this now as it's solved in the 1.0.2 release, thanks.
